sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -211,13 +211,15 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     }
 
     assertNotBucketed("save")
-    val dataSource = DataSource(
-      df.sparkSession,
-      className = source,
-      partitionColumns = partitioningColumns.getOrElse(Nil),
-      options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    runCommand(df.sparkSession, "save") {
+      SaveIntoDataSourceCommand(
```

Comment:
Does this also cover SPARK-19557? If so, might as well mention that in the PR (or close the bug as a duplicate or related or something).